File: crates/holos-tda-py/python/holos_tda/records.py
def _class_result(raw):
    bars, records = raw
    classes = []
    for (
        group_id,
        class_id,
        basis_index,
        birth,
        death,
        modulus,
        scale,
        terms,
        active_graph_digest,
        class_digest,
    ) in records:
        classes.append(
            {
                "group_id": group_id,
                "id": class_id,
                "basis_index": basis_index,
                "birth": birth,
                "death": death,
                "essential": death is None,
                "modulus": modulus,
                "scale": scale,
                "terms": terms,
                "provenance": {
                    "schema": "holos-persistent-class-source-v1",
                    "active_graph_digest": active_graph_digest,
                    "class_digest": class_digest,
                    "interval": (1, birth, death),
                    "birth": birth,
                    "death": death,
                    "modulus": modulus,
                    "scale": scale,
                },
            }
        )
    return bars, classes


def _class_record(record):
    group_id, class_id, basis_index, birth, death, modulus, scale, terms = record
    return {
        "group_id": group_id,
        "id": class_id,
        "basis_index": basis_index,
        "birth": birth,
        "death": death,
        "essential": death is None,
        "modulus": modulus,
        "scale": scale,
        "terms": terms,
    }


def _critical_records(critical):
    return [
        {
            "birth": {"vertices": pair[0], "value": pair[1]},
            "death": None
            if pair[2] is None
            else {"vertices": pair[2][0], "value": pair[2][1]},
        }
        for pair in critical
    ]


def _atlas_result(raw):
    bars, raw_spaces, raw_sensitivities = raw
    spaces = []
    for lineage, group_id, birth, death, basis, critical in raw_spaces:
        spaces.append(
            {
                "lineage": lineage,
                "group_id": group_id,
                "birth": birth,
                "death": death,
                "essential": death is None,
                "multiplicity": len(basis),
                "basis": [_class_record(record) for record in basis],
                "critical_pairs": _critical_records(critical),
            }
        )
    sensitivities = []
    for lineage, birth, death in raw_sensitivities:
        sensitivities.append(
            {
                "lineage": lineage,
                "birth": {"kind": birth[0], "edges": birth[1]},
                "death": {"kind": death[0], "edges": death[1]},
            }
        )
    return {
        "bars": bars,
        "spaces": spaces,
        "sensitivities": sensitivities,
    }


def _event_record(record):
    kind, first, second, old_first, new_first, old_second, new_second = record
    return {
        "kind": kind,
        "first": first,
        "second": second,
        "old_first": old_first,
        "new_first": new_first,
        "old_second": old_second,
        "new_second": new_second,
    }


def _program_result(raw):
    bars, raw_spaces = raw
    spaces = []
    for group_id, birth, death, basis, critical in raw_spaces:
        spaces.append(
            {
                "group_id": group_id,
                "birth": birth,
                "death": death,
                "essential": death is None,
                "multiplicity": len(basis),
                "basis": [_class_record(record) for record in basis],
                "critical_pairs": _critical_records(critical),
            }
        )
    return {"bars": bars, "spaces": spaces}


def _program_work(raw):
    names = (
        "edges_checked",
        "h0_edges_scanned",
        "guards_checked",
        "atoms_touched",
        "atoms_reused",
        "atoms_repaired",
        "atoms_rebuilt",
        "reduction_columns_reused",
        "reduction_columns_reduced",
        "reduction_column_additions",
    )
    return dict(zip(names, raw))


def _program_event(raw):
    kind, atom, edge, guard = raw
    return {"kind": kind, "atom": atom, "edge": edge, "guard": guard}


def _continuation(raw):
    kind, old_spaces, new_spaces, transport = raw
    return {
        "kind": kind,
        "old_spaces": old_spaces,
        "new_spaces": new_spaces,
        "transport": [
            {"old": old, "new": new, "coefficient": coefficient}
            for old, new, coefficient in transport
        ],
    }


def _correspondence(raw):
    (
        old_space,
        new_space,
        scale,
        old_rank,
        new_rank,
        old_image_rank,
        new_image_rank,
        relation_rank,
        basis,
    ) = raw
    return {
        "old_space": old_space,
        "new_space": new_space,
        "scale": scale,
        "old_rank": old_rank,
        "new_rank": new_rank,
        "old_image_rank": old_image_rank,
        "new_image_rank": new_image_rank,
        "relation_rank": relation_rank,
        "is_isomorphism": (
            relation_rank == old_rank == new_rank
            and old_image_rank == old_rank
            and new_image_rank == new_rank
        ),
        "basis": [
            {
                "old": [
                    {"basis": item[0], "coefficient": item[1]} for item in vector[0]
                ],
                "new": [
                    {"basis": item[0], "coefficient": item[1]} for item in vector[1]
                ],
            }
            for vector in basis
        ],
    }


def _program_update(raw):
    mode, events, continuation, correspondence, work, result = raw
    return {
        "mode": mode,
        "events": [_program_event(record) for record in events],
        "continuation": [_continuation(record) for record in continuation],
        "correspondence": [_correspondence(record) for record in correspondence],
        "work": _program_work(work),
        "result": _program_result(result),
    }


def _index_work(raw):
    names = (
        "edges_checked",
        "nodes_touched",
        "nodes_shared",
        "nodes_repaired",
        "nodes_rebuilt",
        "nodes_composed",
        "relative_nodes_rebuilt",
        "relative_nodes_composed",
        "relative_input_cells",
        "relative_core_cells",
        "relative_cancellations",
        "reduction_columns_reused",
        "reduction_columns_reduced",
        "reduction_column_additions",
    )
    raw = list(raw)
    relative = list(raw.pop(6))
    return dict(zip(names, raw[:6] + relative + raw[6:]))


def _index_event(raw):
    kind, node, edge = raw
    return {"kind": kind, "node": node, "edge": edge}


def _index_update(raw):
    (
        mode,
        bars,
        removed,
        added,
        events,
        correspondence,
        work,
        version,
        delta_proof,
        snapshot_proof,
    ) = raw
    return {
        "mode": mode,
        "bars": bars,
        "diagram_delta": {"removed": removed, "added": added},
        "events": [_index_event(record) for record in events],
        "correspondence": [_correspondence(record) for record in correspondence],
        "work": _index_work(work),
        "version": version,
        "delta_proof": (None if delta_proof is None else bytes(delta_proof)),
        "snapshot_proof": (None if snapshot_proof is None else bytes(snapshot_proof)),
    }


def _intervention(raw):
    status, target, lower, upper, edits, result, artifact = raw
    return {
        "status": status,
        "target": target,
        "lower_bound": lower,
        "upper_bound": upper,
        "edits": [
            {"edge": edge, "before": before, "after": after}
            for edge, before, after in edits
        ],
        "result": None if result is None else _program_result(result),
        "artifact": None if artifact is None else bytes(artifact),
    }

File: crates/holos-tda-py/src/common/records.rs
//! Shared Python record conversions.

use std::fmt::Write as _;

use pyo3::prelude::PyResult;

use super::config::display_err;
use super::core::*;
use super::types::*;

/// pyo3 converts essential death `f64::INFINITY` to `math.inf`.
pub(crate) fn to_bars(mut diagram: holos_tda::Diagram) -> Bars {
    diagram.canonicalize();
    diagram
        .bars
        .into_iter()
        .map(|b| (b.dim, b.birth, b.death))
        .collect()
}

pub(crate) fn to_explained(explained: ExplainedDiagram) -> holos_tda::Result<Explained> {
    let bars = to_bars(explained.diagram);
    let classes = explained
        .spaces
        .into_iter()
        .flat_map(|space| space.basis)
        .map(|class| {
            let provenance = class.provenance.as_ref().ok_or_else(|| {
                holos_tda::Error::InvalidInput(
                    "scalar persistence class has no interval-bound provenance".into(),
                )
            })?;
            let death = (!class.interval.is_essential()).then_some(class.interval.death);
            let terms = class
                .cocycle
                .terms
                .into_iter()
                .map(|term| (term.u, term.v, term.coefficient))
                .collect();
            Ok((
                class.group_id.to_string(),
                class.id.to_string(),
                class.basis_index,
                class.interval.birth,
                death,
                class.cocycle.modulus,
                class.cocycle.scale,
                terms,
                hex_digest(provenance.source_graph_digest()),
                hex_digest(provenance.class_digest()),
            ))
        })
        .collect::<holos_tda::Result<Vec<_>>>()?;
    Ok((bars, classes))
}

fn hex_digest(digest: &[u8; 32]) -> String {
    let mut result = String::with_capacity(64);
    for byte in digest {
        write!(&mut result, "{byte:02x}").expect("writing to a String cannot fail");
    }
    result
}

pub(crate) fn class_record(class: holos_tda::PersistentClass) -> ClassRecord {
    let death = (!class.interval.is_essential()).then_some(class.interval.death);
    let terms = class
        .cocycle
        .terms
        .into_iter()
        .map(|term| (term.u, term.v, term.coefficient))
        .collect();
    (
        class.group_id.to_string(),
        class.id.to_string(),
        class.basis_index,
        class.interval.birth,
        death,
        class.cocycle.modulus,
        class.cocycle.scale,
        terms,
    )
}

pub(crate) fn gradient_record(gradient: EndpointGradient) -> GradientRecord {
    match gradient {
        EndpointGradient::Edge(edge) => ("edge".into(), vec![(edge.u, edge.v)]),
        EndpointGradient::Tied(edges) => (
            "tied".into(),
            edges.into_iter().map(|edge| (edge.u, edge.v)).collect(),
        ),
        EndpointGradient::Essential => ("essential".into(), Vec::new()),
        _ => ("unknown".into(), Vec::new()),
    }
}

pub(crate) fn to_atlas_result(evaluation: AtlasEvaluation) -> AtlasResult {
    let bars = to_bars(evaluation.diagram);
    let spaces = evaluation
        .spaces
        .into_iter()
        .map(|evaluated| {
            let space = evaluated.space;
            let death = (!space.interval.is_essential()).then_some(space.interval.death);
            let basis = space.basis.into_iter().map(class_record).collect();
            let critical = space
                .critical_pairs
                .into_iter()
                .map(|pair| {
                    (
                        pair.birth.vertices,
                        pair.birth.value,
                        pair.death.map(|death| (death.vertices, death.value)),
                    )
                })
                .collect();
            (
                evaluated.lineage.to_string(),
                space.id.to_string(),
                space.interval.birth,
                death,
                basis,
                critical,
            )
        })
        .collect();
    let sensitivities = evaluation
        .sensitivities
        .into_iter()
        .map(|sensitivity| {
            (
                sensitivity.lineage.to_string(),
                gradient_record(sensitivity.birth),
                gradient_record(sensitivity.death),
            )
        })
        .collect();
    (bars, spaces, sensitivities)
}

pub(crate) fn to_program_result(explained: ExplainedDiagram) -> ProgramResult {
    let bars = to_bars(explained.diagram);
    let spaces = explained
        .spaces
        .into_iter()
        .map(|space| {
            let death = (!space.interval.is_essential()).then_some(space.interval.death);
            let basis = space.basis.into_iter().map(class_record).collect();
            let critical = space
                .critical_pairs
                .into_iter()
                .map(|pair| {
                    (
                        pair.birth.vertices,
                        pair.birth.value,
                        pair.death.map(|death| (death.vertices, death.value)),
                    )
                })
                .collect();
            (
                space.id.to_string(),
                space.interval.birth,
                death,
                basis,
                critical,
            )
        })
        .collect();
    (bars, spaces)
}

pub(crate) fn work_record(work: ProgramWork) -> WorkRecord {
    (
        work.edges_checked,
        work.h0_edges_scanned,
        work.guards_checked,
        work.atoms_touched,
        work.atoms_reused,
        work.atoms_repaired,
        work.atoms_rebuilt,
        work.reduction_columns_reused,
        work.reduction_columns_reduced,
        work.reduction_column_additions,
    )
}

pub(crate) fn program_event_kind(kind: ProgramEventKind) -> &'static str {
    match kind {
        ProgramEventKind::VertexSetChanged => "vertex_set_changed",
        ProgramEventKind::EdgeSetChanged => "edge_set_changed",
        ProgramEventKind::ThresholdCrossing => "threshold_crossing",
        ProgramEventKind::GuardFailed => "guard_failed",
        ProgramEventKind::AtomRebuilt => "atom_rebuilt",
        ProgramEventKind::ReductionSuffixRepaired => "reduction_suffix_repaired",
        ProgramEventKind::SeparatorContractChanged => "separator_contract_changed",
        _ => "unknown",
    }
}

pub(crate) fn program_event_record(event: ProgramEvent) -> ProgramEventRecord {
    (
        program_event_kind(event.kind).into(),
        event.atom,
        event.edge.map(|edge| (edge.u, edge.v)),
        event.guard.map(|guard| match guard {
            holos_tda::ReductionGuardKind::ChangeOfBasis => "change_of_basis".into(),
            holos_tda::ReductionGuardKind::Pivot => "pivot".into(),
            _ => "unknown".into(),
        }),
    )
}

pub(crate) fn continuation_kind(kind: holos_tda::ContinuationKind) -> &'static str {
    match kind {
        holos_tda::ContinuationKind::Isomorphism => "isomorphism",
        holos_tda::ContinuationKind::Split => "split",
        holos_tda::ContinuationKind::Merge => "merge",
        holos_tda::ContinuationKind::Mixing => "mixing",
        holos_tda::ContinuationKind::Birth => "birth",
        holos_tda::ContinuationKind::Death => "death",
        holos_tda::ContinuationKind::Ambiguous => "ambiguous",
        _ => "unknown",
    }
}

pub(crate) fn continuation_record(value: holos_tda::ClassContinuation) -> ContinuationRecord {
    (
        continuation_kind(value.kind).into(),
        value
            .old_spaces
            .into_iter()
            .map(|space| space.to_string())
            .collect(),
        value
            .new_spaces
            .into_iter()
            .map(|space| space.to_string())
            .collect(),
        value
            .transport
            .into_iter()
            .map(|item| (item.old.to_string(), item.new.to_string(), item.coefficient))
            .collect(),
    )
}

pub(crate) fn correspondence_record(value: holos_tda::ClassCorrespondence) -> CorrespondenceRecord {
    (
        value.old_space.to_string(),
        value.new_space.to_string(),
        value.scale,
        value.old_rank,
        value.new_rank,
        value.old_image_rank,
        value.new_image_rank,
        value.relation_rank,
        value
            .basis
            .into_iter()
            .map(|vector| {
                (
                    vector
                        .old
                        .into_iter()
                        .map(|term| (term.basis.to_string(), term.coefficient))
                        .collect(),
                    vector
                        .new
                        .into_iter()
                        .map(|term| (term.basis.to_string(), term.coefficient))
                        .collect(),
                )
            })
            .collect(),
    )
}

pub(crate) fn program_mode(mode: ProgramUpdateMode) -> &'static str {
    match mode {
        ProgramUpdateMode::Reused => "reused",
        ProgramUpdateMode::Repaired => "repaired",
        ProgramUpdateMode::Recompiled => "recompiled",
    }
}

pub(crate) fn program_update_record(update: ProgramUpdate) -> ProgramUpdateRecord {
    (
        program_mode(update.mode).into(),
        update
            .events
            .into_iter()
            .map(program_event_record)
            .collect(),
        update
            .continuation
            .into_iter()
            .map(continuation_record)
            .collect(),
        update
            .correspondence
            .into_iter()
            .map(correspondence_record)
            .collect(),
        work_record(update.work),
        to_program_result(update.result),
    )
}

pub(crate) fn digest_string(digest: &[u8; 32]) -> String {
    let mut output = String::with_capacity(64);
    for byte in digest {
        write!(&mut output, "{byte:02x}").expect("writing to a string cannot fail");
    }
    output
}

pub(crate) fn bars_from(bars: Vec<holos_tda::Bar>) -> Bars {
    bars.into_iter()
        .map(|bar| (bar.dim, bar.birth, bar.death))
        .collect()
}

pub(crate) fn index_mode(mode: IndexUpdateMode) -> &'static str {
    match mode {
        IndexUpdateMode::Unchanged => "unchanged",
        IndexUpdateMode::Repaired => "repaired",
        IndexUpdateMode::Composed => "composed",
        IndexUpdateMode::Relative => "relative",
        IndexUpdateMode::Rebuilt => "rebuilt",
        IndexUpdateMode::Recompiled => "recompiled",
    }
}

pub(crate) fn index_event_kind(kind: IndexEventKind) -> &'static str {
    match kind {
        IndexEventKind::ThresholdCrossing => "threshold_crossing",
        IndexEventKind::ReductionRepaired => "reduction_repaired",
        IndexEventKind::ReductionRebuilt => "reduction_rebuilt",
        IndexEventKind::InterfaceComposed => "interface_composed",
        IndexEventKind::RelativeCoreRebuilt => "relative_core_rebuilt",
        IndexEventKind::RelativeCoreComposed => "relative_core_composed",
        IndexEventKind::EnvelopeRecompiled => "envelope_recompiled",
        _ => "unknown",
    }
}

pub(crate) fn index_event_record(event: IndexEvent) -> IndexEventRecord {
    (
        index_event_kind(event.kind).into(),
        event.node.map(|digest| digest_string(&digest)),
        event.edge.map(|edge| (edge.u, edge.v)),
    )
}

pub(crate) fn index_work_record(work: IndexWork) -> IndexWorkRecord {
    (
        work.edges_checked,
        work.nodes_touched,
        work.nodes_shared,
        work.nodes_repaired,
        work.nodes_rebuilt,
        work.nodes_composed,
        (
            work.relative_nodes_rebuilt,
            work.relative_nodes_composed,
            work.relative_input_cells,
            work.relative_core_cells,
            work.relative_cancellations,
        ),
        work.reduction_columns_reused,
        work.reduction_columns_reduced,
        work.reduction_column_additions,
    )
}

pub(crate) fn index_update_record(
    old: &PersistenceIndex,
    transition: &IndexTransition,
) -> PyResult<IndexUpdateRecord> {
    let (delta_proof, snapshot_proof) = if transition.mode == IndexUpdateMode::Recompiled {
        let proof = IndexSnapshotProof::from_index(&transition.index).map_err(display_err)?;
        (None, Some(proof.encode().map_err(display_err)?))
    } else {
        let proof = IndexDeltaProof::between(old, &transition.index).map_err(display_err)?;
        (Some(proof.encode().map_err(display_err)?), None)
    };
    Ok((
        index_mode(transition.mode).into(),
        to_bars(transition.index.diagram().clone()),
        bars_from(transition.delta.removed.clone()),
        bars_from(transition.delta.added.clone()),
        transition
            .events
            .clone()
            .into_iter()
            .map(index_event_record)
            .collect(),
        transition
            .correspondence
            .clone()
            .into_iter()
            .map(correspondence_record)
            .collect(),
        index_work_record(transition.work),
        digest_string(&transition.index.version()),
        delta_proof,
        snapshot_proof,
    ))
}

pub(crate) fn program_decode_limits() -> ProgramDecodeLimits {
    ProgramDecodeLimits::default()
}

pub(crate) fn program_trace_decode_limits() -> ProgramTraceDecodeLimits {
    ProgramTraceDecodeLimits::default()
}

pub(crate) fn event_kind(kind: TopologyEventKind) -> String {
    match kind {
        TopologyEventKind::VertexSetChanged => "vertex_set_changed",
        TopologyEventKind::EdgeSetChanged => "edge_set_changed",
        TopologyEventKind::ThresholdCrossing => "threshold_crossing",
        TopologyEventKind::EqualitySplit => "equality_split",
        TopologyEventKind::EqualityMerge => "equality_merge",
        TopologyEventKind::OrderSwap => "order_swap",
        _ => "unknown",
    }
    .into()
}

pub(crate) fn event_record(event: TopologyEvent) -> EventRecord {
    (
        event_kind(event.kind),
        event.first.map(|edge| (edge.u, edge.v)),
        event.second.map(|edge| (edge.u, edge.v)),
        event.old_first,
        event.new_first,
        event.old_second,
        event.new_second,
    )
}

pub(crate) fn point_gradient_record(gradient: PointEndpointGradient) -> PointGradientValue {
    (
        (gradient.edge.u, gradient.edge.v),
        gradient
            .terms
            .into_iter()
            .map(|term| (term.point, term.coordinate, term.value))
            .collect(),
    )
}

pub(crate) fn point_sensitivity_records(
    atlas: &PointPersistenceAtlas,
) -> Vec<PointSensitivityRecord> {
    atlas
        .sensitivities()
        .into_iter()
        .map(|sensitivity| {
            (
                sensitivity.lineage.to_string(),
                sensitivity.birth.map(point_gradient_record),
                sensitivity.death.map(point_gradient_record),
            )
        })
        .collect()
}
